common/djangoapps/enrollment/views.py

```diff
@@ -345,12 +369,25 @@ def post(self, request):
             )
 
         try:
-            response = api.add_enrollment(user, unicode(course_id))
+            # Check if the user is currently enrolled, and if it is the same as the current enrolled mode. We do not
+            # have to check if it is inactive or not, because if it is, we are still upgrading if the mode is different,
+            # and either path will re-activate the enrollment.
+            #
+            # Only server-to-server calls will currently be allowed to modify the mode for existing enrollments. All
+            # other requests will go through add_enrollment(), which will allow creating of new enrollments, and
+            # re-activating enrollments
+            enrollment = api.get_enrollment(user, unicode(course_id))
+            if has_api_key_permissions and enrollment and enrollment['mode'] != mode:
+                response = api.update_enrollment(user, unicode(course_id), mode=mode)
+                http_success_status = status.HTTP_200_OK
+            else:
+                response = api.add_enrollment(user, unicode(course_id), mode=mode)
```

Comment:
This will create an enrollment if `enrollment['mode'] == mode`. Try nesting to avoid the unnecessary call:

```python
if has_api_key_permissions and enrollment:
  if enrollment['mode'] != mode:
    # update
    ...
  else:
    # NO-OP (only here for completeness)
else:
  # add
  ...
```